### ISSUE TITLE
Allow extended parser models to pass mypy

### ISSUE THREAD

**Issue by ClaytonJY** [awslabs/aws-lambda-powertools-python#849]
**Is your feature request related to a problem? Please describe.**

The docs example for extending `parser` models (https://awslabs.github.io/aws-lambda-powertools-python/latest/utilities/parser/#extending-built-in-models) doesn't pass `mypy`, because the `Order` class is not compatible with the `EventBridgeModel` annotation `detail: Dict[str, Any]`.
```bash
error: Incompatible types in assignment (expression has type "Order", base class "EventBridgeModel" defined the type as "Dict[str, Any]")
```
I would like to be able to both extend this model and have my code pass type checks without `# type: ignore` or similar.

I suspect this affects other `parser` models, like the `Message` field of `SnsNotificationModel`.

**Describe the solution you'd like**

One solution is to change only the `detail` type annotation, e.g. `Union[Dict[str, Any], BaseModel]`, but there may be other things to add to that list and it doesn't feel great; there could be many other types that would parse fine and could be listed there, right? Or is that it, because you either use `EventBridgeModel` directly (`Dict`) or you subclass it (`BaseModel`)?

Is there anything that could be done with `typing.Protocol`, to check the class has certain methods regardless of name?

There might also be an argument to not change the type, but note in the docs that this will fail type-checks and instruct how to silence `mypy` in such instances.

**Describe alternatives you've considered**

I tried various solutions which didn't work, like adding `dict` or `TypedDict` to the parent class list for `Order` (incompatible methods), or changing it to `TypedDict` (same error as above).

**Additional context**

I'm using the `pydantic.mypy` extension.

**Comment by heitorlessa** [comment]
hey @ClaytonJY thanks a lot for flagging this. I'm not entirely sure how to solve this yet, we'd have to maybe experiment with a generic type bound to BaseModel.

cc @ran-isenberg in case he knows as this should be a fairly common problem in Pydantic world

**Comment by ran-isenberg** [comment]
I didn't encounter this issue yet to be honest.

**Comment by ClaytonJY** [comment]
FWIW I'm a very new user of this package, and since posting this have come to avoid the issue a different way: use the envelope and never bother subclassing that model. Still, given the benefits of this package around typing, it would be nice if examples in the docs like this were type compliant.

Happy to help, but may need to be pointed in the right direction. Very curious to see where this discussion goes, and thank you for a package that has cut at least half of my code just a few hours after discovering it 🙏

**Comment by ran-isenberg** [comment]
> FWIW I'm a very new user of this package, and since posting this have come to avoid the issue a different way: use the envelope and never bother subclassing that model. Still, given the benefits of this package around typing, it would be nice if examples in the docs like this were type compliant.
> 
> Happy to help, but may need to be pointed in the right direction. Very curious to see where this discussion goes, and thank you for a package that has cut at least half of my code just a few hours after discovering it 🙏
@ClaytonJY 
There are many advantages to subclassing the Model and not using the envelope - especially when you want both the body and the meta data arguments from the subclass.
In that case, when i subclass I just set the body/details argument to my actual real BaseModel class (MyBusinessLogic class etc.) and create a validator(pre=True) that takes the str/dict and does a json.loads(details). That way I get both the parsed metadata class and my business logic class parsed correctly. I can supply a code example if that helps.

If you dont care about the metadata arguments, just use envelope and model

**Comment by ran-isenberg** [comment]
> FWIW I'm a very new user of this package, and since posting this have come to avoid the issue a different way: use the envelope and never bother subclassing that model. Still, given the benefits of this package around typing, it would be nice if examples in the docs like this were type compliant.
> 
> Happy to help, but may need to be pointed in the right direction. Very curious to see where this discussion goes, and thank you for a package that has cut at least half of my code just a few hours after discovering it 🙏

Subclassing has a lot of advantages -if you require both metadata and business logic (body/details) attributes. 

I guess mypy doesnt like that we change the type of the baseclass attribute. I'm not sure how you can avoid that, that's the whole point of the parsing trick we do there.

**Comment by heitorlessa** [comment]
Hey! Once we address two P1 issues (Batch, Idempotency), I’ll try replicate
with mypy.

Clayton - could you share what mypy config you’re using when you have a
chance?

In theory, subclassing shouldn’t cause this issue with typing, as you’re
replacing it. If it indeed does and there isn’t an elegant solution, happy
to make a generic for all data fields (detail, body, etc) where you could
pass a str or a type bounded to BaseModel.

**Comment by heitorlessa** [comment]
One thing I might have missed to ask is whether you’re using Pydantic mypy
plugin, otherwise Mypy could indeed complain of the type change when
subclassing - forgot to mention that in the docs

https://pydantic-docs.helpmanual.io/mypy_plugin/

**Comment by ClaytonJY** [comment]
@heitorlessa I no longer have access to that exact project where I discovered this, but I'm fairly certain the only mypy config I had set was using the pydantic plugin. I may have fed some command-line args to mypy like ignoring un-annotated imports. I'm quite certain I wasn't doing anything exotic!

I could whip up a minimal reprex if needed, would that help?

**Comment by heitorlessa** [comment]
Yes please! That would speed things up on our end when we get to it this
week

**Comment by ClaytonJY** [comment]
@heitorlessa here's that reprex, as a new repo: https://github.com/ClaytonJY/powertools-type-error-example

thanks for caring!

**Comment by heitorlessa** [comment]
Looking at this today @ClaytonJY thank you so much again for creating the repro, it makes a lot easier for us :)

**Comment by heitorlessa** [comment]
As expected, a generic type bound to BaseModel in an Union makes MyPy happy. If the actual payload is passed and no override happens, it correctly detects `detail` as a dictionary.

It'd be impractical to allow any other field to be overridden by type, so I'll go ahead and ensure all built-in Models can be overridden on fields where the payload is (e.g. `detail` in EventBridge, `body` in API Gateway, etc.)

**Comment by heitorlessa** [comment]
PR available - should be available in the next release: https://github.com/awslabs/aws-lambda-powertools-python/pull/883

**Comment by heitorlessa** [comment]
Now released as part of 1.23 version
